### PR TITLE
Kitchen CI test for verify enclave

### PR DESCRIPTION
**Pull request by DavidJeche**
# Why I am making this change

I am making this change in order to add infrastructure test for the verify environment. We adding in a number of different tools that we hope to improve our workflow. What we have added:

- Test Kitchen - The harness for running test / https://kitchen.ci/
- Kitchen Terraform - Terraform plugin for kitchen / https://newcontext-oss.github.io/kitchen-terraform/
- Inspec - This is the verification system used to run test / https://www.inspec.io/

The process is fairly simple. You can go into the the folder `terraform/modules/enclave/prometheus`. Once in this folder you can then proceed to run a `kitchen converge` this should generate a development stack that you can use for running the test or developing the stack. 

The output provided after a successful converge will show the DNS name use can use to log in to the insatnce. You can run `kitchen veify` this will allow you to run allow tests againest the instance.

# What approach I took

I introduced a set of new workflows that we can try out in our development workflow. This should help write better code, simplify development and imrpove pull request review time. We have everything in a self containered kitchen file. The tests are written in ruby and provide resources that have great documention that can make usage of the tool very simple.

## Merge after Ed's merge

### CODE REVIEW

File: terraform/modules/enclave/prometheus/test/integration/prometheus/controls/aws_cloud_resources.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  describe aws_s3_bucket(bucket_name: s3_bucket_id) do
+    it { should exist }
+    it { should_not be_public }
+  end
+
+  describe aws_security_group(group_name: 'observe-hub-perf-test-prometheus-instance') do
+    it { should exist }
+    its('group_name') { should eq 'observe-hub-perf-test-prometheus-instance' }
+    it { should allow_out(port: '9100', ipv4_range: '10.0.0.0/22') }
+    it { should allow_in(port: '22', ipv4_range: allow_ip_subnets) }
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/prometheus/test/integration/prometheus/controls/aws_cloud_resources.rb]: Is there a way to make sure those only lets in this ipv4 range and not a larger one?

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf [hunk 2]
```diff
@@ -0,0 +1,75 @@
+provider "aws" {
+  region = "eu-west-2" # "${var.aws_region}"
+}
+
+terraform {
+  required_version = ">= 0.11.7"
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/prometheus/test/fixtures/main.tf]: We might want to pin this exactly to 0.11.7 like we do for all the rest of the codebase

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  enable_dns_support   = true
+}
+
+resource "aws_internet_gateway" "internet_gateway" {
+  vpc_id = "${aws_vpc.main.id}"
+}
+
+module "network" {
+  source              = "../../../../enclave/network"
+  environment         = "perf-test"
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/prometheus/test/fixtures/main.tf]: Might be good to call this `verify-perf-a-test` to match the name of the project

Comment (idavidmcdonald) [reply terraform/modules/enclave/prometheus/test/fixtures/main.tf]: Also will we run into issues if more than one person tries to run the tests at the same time? (although this probably is something we could survive with for the moment)

File: terraform/modules/enclave/prometheus/test/integration/prometheus/controls/operanting_systems.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+control "operating_system" do
+  
+  describe os.family do
+    it { should eq 'debian' }
+  end
+
+  describe file('/etc/cron.d/config_pull') do
+    its('owner') { should eq 'root' }
+    its('group') { should eq 'root' }
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/prometheus/test/integration/prometheus/controls/operanting_systems.rb]: Do we want to check mode on this too?

File: terraform/modules/enclave/prometheus/test/integration/prometheus/controls/operanting_systems.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
+control "operating_system" do
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/prometheus/test/integration/prometheus/controls/operanting_systems.rb]: Just spotted that this file is misspelt as `operanting_systems.rb` as opposed to `operating_systems.rb`

File: .gitignore [hunk 6]
```diff
@@ -18,3 +18,11 @@ environment.sh
 
 # os files 
 .DS_Store
+
+
+.kitchen/**
+.kitchen/logs/kitchen.log
```
[the remark this answers is not in the document]
Comment (DavidJeche) [reply .gitignore]: This is based on things that we had when rebasing. Some stuff had made it into the rebased content.

File: .gitignore [hunk 7]
```diff
@@ -18,3 +18,11 @@ environment.sh
...
+terraform/modules/enclave/prometheus/.kitchen/
```
[the remark this answers is not in the document]
Comment (DavidJeche) [reply .gitignore]: Removed this and a few of the ones beginning taken care of by the regex.

File: terraform/modules/enclave/prometheus/Gemfile [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/Gemfile]: Yep nice catch 👍

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  enable_ssh = true
+
+  product     = "hub"
+  environment = "perf-test"
+
+  subnet_ids          = "${module.network.subnet_ids}"
+  availability_zones  = "${module.network.availability_zones}"
+  vpc_security_groups = ["${module.network.security_groups}", "${aws_security_group.permit_internet_access.id}"]
+  ec2_endpoint_ips    = "${module.network.endpoint_network_interface_ip}"
+  verify_enclave      = "false"
```
[the remark this answers is not in the document]
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/test/fixtures/main.tf]: Ok removed these

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    from_port = 443
+    to_port   = 443
+
+    cidr_blocks = [
+      "0.0.0.0/0",
+    ]
+  }
+
+  tags {
+    Name = "SSH from GDS"
```
[the remark this answers is not in the document]
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/test/fixtures/main.tf]: I have replaced this and added the ability to access the Prometheus for dev environments only.

File: terraform/modules/enclave/prometheus/test/fixtures/variables.tf [hunk 11]
```diff
@@ -0,0 +1,11 @@
+variable "az_zone" {
```
[the remark this answers is not in the document]
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/test/fixtures/variables.tf]: The az_zone we do not need anymore but the endpoint IP's is a good one to have as a placeholder.

File: .gitignore [hunk 12]
```diff
@@ -18,3 +18,11 @@ environment.sh
...
+*.plan
+*/.kitchen/**
+*.lock
```
Comment (idavidmcdonald) [inline .gitignore]: Looks like we should remove this line actually as per https://stackoverflow.com/questions/4151495/should-gemfile-lock-be-included-in-gitignore

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf (hunk 3 above)
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/test/fixtures/main.tf]: Yep for the time being it is basic and this means that we can use this until we get a make file. The reason why I avoided that was that I did not want to run in to charecter limits.

File: .gitignore [hunk 13]
```diff
@@ -18,3 +18,7 @@ environment.sh
 
 # os files 
 .DS_Store
+
+*.plan
+*/.kitchen/**
+*.lock
```
Comment (idavidmcdonald) [inline .gitignore]: https://stackoverflow.com/questions/4151495/should-gemfile-lock-be-included-in-gitignore

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    from_port = 9090
+    to_port   = 9090
+
+    cidr_blocks = [
+      "0.0.0.0/0",
+    ]
+  }
+
+  tags {
+    Name = "SSH & prometheus access from GDS in dev env"
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/prometheus/test/fixtures/main.tf]: Is this actually SSH access though? I see no port 22

File: terraform/modules/enclave/prometheus/test/integration/prometheus/controls/aws_cloud_resources.rb [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  describe aws_iam_role('prometheus_profile') do
+    it { should exist }
+  end
+
+  describe aws_iam_policy('prometheus_instance_profile') do
+    it { should exist }
+    its('attached_roles') { should include "prometheus_profile" }
+    it { should be_attached }
+    it { should have_statement(Action: ['s3:Get*','s3:ListBucket'], Effect: 'Allow', Sid: 's3Bucket') }
+    it { should have_statement(Action: 'ec2:Describe*', Effect: 'Allow', Resource: '*', Sid: 'ec2Policy') }
```
[the remark this answers is not in the document]
Comment (idavidmcdonald) [reply terraform/modules/enclave/prometheus/test/integration/prometheus/controls/aws_cloud_resources.rb]: Maybe the `statement_count` variable could be useful? https://www.inspec.io/docs/reference/resources/aws_iam_policy/

File: terraform/modules/enclave/prometheus/test/fixtures/main.tf (hunk 14 above)
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/test/fixtures/main.tf]: It is supposed to be internet access as by default the node can't talk to the outside world.

File: terraform/modules/enclave/README.md [hunk 16]
```diff
@@ -0,0 +1,33 @@
+# Prometheus verify enclave
+
+The enclave environment is reference to the area of verify that we use to perform  our deployment. This environment has a number of different security requirements that difficult and a very custom environment to deploy too.
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/README.md]: I'm not sure if the second sentence here is adding much value so maybe could be either deleted or expanded on to be specific (although feel free to disagree). I think there is a missing word too before `difficult`.

File: terraform/modules/enclave/prometheus/test/integration/prometheus/controls/aws_cloud_resources.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (DavidJeche) [reply terraform/modules/enclave/prometheus/test/integration/prometheus/controls/aws_cloud_resources.rb]: I had not come across this statement count. I have added it in now. We tried to get the most basic functionality then improve on them as time progressed.

File: terraform/modules/enclave/README.md [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+3. `aws-vault exec <created profile> -- kitchen <action>` This is the general command that you can use in order to run the environment. 
+4. `create` `converge` `verify` these are the three commands that can used in order to spin up a stack and test. The converge can be executed multiple times to test changes.
+5. Once you are done developing, testing and using the stack you can then use the action `destroy` in order to bring down the whole stack.
+ 
+## Deploying
+When performing a deployment you need to ensure you have the correct permission to do so as extra permission are required. Please ask a member of the team for more information.
+
+You need to be at the root of the repository. This is where the deployment script is located. It is name `deploy_enclave.sh`. 
+
+The general structure of the deployment command is the following: `/deploy_enclave.sh -e <enviroment> -p <aws vault profile> -a <terraform method> -s <state>`
```
Comment (idavidmcdonald) [inline terraform/modules/enclave/README.md]: Optional: Add an example of this command

**Pull request merged**